enhancements/olm/safe-cluster-upgrades.md

```diff
@@ -0,0 +1,267 @@
+
+When OLM is reconciling CSV it will check the set of annotations for an annotation whose key matches `operators.coreos.com/maxOpenShiftVersion` and the value is the maximum version of OpenShift that the operators supports. An example of this annotation can be seen below:
+
+```yaml=
+apiVersion: operators.coreos.com/v1alpha1
+kind: ClusterServiceVersion
+metadata:
+  annotations:
+    # Prevent upgrades to OpenShift Version 4.9
+    operators.coreos.com/maxOpenShiftVersion: 4.8.0 
+```
+
+> NOTE: OLM should allow users to specify Major.Minor.Patch versions in this field BUT OLM can only prevent Major.Minor OpenShift upgrades. By allowing Major.Minor.Patch versions to be specified, OLM will not need to modify the existing annotations if OpenShift ever allows ClusterOperators to prevent patch upgrades.
+
+Assume that the CSV shown above is the only CSV on the Cluster. In this case, OLM would compare the value of the `operators.coreos.com/maxOpenShiftVersion` annotation against the version of the Cluster to decide it's upgradeable status.
+
+Examples can be found below:
+
+| OpenShift Cluster Version   | CSV's Max OpenShift Version | Is Upgradeable?|
+| --------------------------- | --------------------------- | -------------- |
+| 4.6.z                       | 4.8.0                       | True           |
+| 4.7.z                       | 4.8.0                       | True           |
+| 4.8.z                       | 4.8.0                       | False          |
+
+Examples of possible OLM Upgradeable conditions based on installed operators can be found below:
+```yaml=
+
+```
+
+As this is an opt-in feature, there are instances where OLM will be unable to determine if the upgrade is safe:
+
+- The CSV is missing the `operators.coreos.com/maxOpenShiftVersion` annotation
+- The value associated with the `operators.coreos.com/maxOpenShiftVersion` is not a valid Semantic Version (Valid Example: x.y.z).
+
+OLM will report the following upgradeable conditions based on the collection of CSV present on cluster:
+
+| Key                | Meaning                                                                                            |
+| -------------------| -------------------------------------------------------------------------------------------------- |
+| Upgradeable    CSV | A CSV with a valid upgradeable annotation that will run on the next Minor version of OpenShift     |
+| Unupgradeable  CSV | A CSV with a valid upgradeable annotation that will not run on the next Minor version of OpenShift |
+| Undeterminable CSV | A CSV that does not include a valid upgradeable annotation                                         |
```

Comment:
how will these conditions be presented to the admin?  In particular the "indeterminate" state needs to be clear to the admin since we are not(?) going to block upgrades when we are in that state.

Reply:
There are two ways to initiate a Cluster Upgrade, via the CLI and OpenShift Console UI.

# CLI
I am unsure how we could notify users that there is an operator that is "indeterminate". I believe the Cluster Admin would need to look at  the OLM ClusterOperator.

# OpenShift Console UI
The messages in ClusterOperators t is already available in the Cluster Admin panel:
![ClusterOperators](https://user-images.githubusercontent.com/5481178/105378136-80b04a80-5bd9-11eb-8bba-b0ac65cf18f2.png)

 The UI Team could work to make this more visible.

Reply:
yeah i don't think there's much we can do about the cli.  Which is all the more reason we need to ensure that if we are in that state, the warning is very clearly presented to the admin.  (We can also document as part of the upgrade process that we recommend you check the OLM status before starting an upgrade).

For the UI, I was thinking we might want an explicit alert when we are in this state, rather than relying on the user to go look at the operator status.  But perhaps that is too aggressive/noisy.

Reply:
Another option for the UI is to present/warn the user when they are on the upgrade panel, that OLM is not happy.

Reply:
> For the UI, I was thinking we might want an explicit alert when we are in this state, rather than relying on the user to go look at the operator status. But perhaps that is too aggressive/noisy.

By "alert" do you mean a warning box or a prometheus alert? I don't know that we need this information picked up by telemeter.

> Another option for the UI is to present/warn the user when they are on the upgrade panel, that OLM is not happy.

I think this is the best option. We could define a specific `reason` that alerts the UI to show a warning box, perhaps @spadgett or someone from his team can chime in.

Reply:
> By "alert" do you mean a warning box or a prometheus alert? I don't know that we need this information picked up by telemeter.

i meant a prometheus alert since those get presented to the admin on the console.  i don't know that we care that it will be picked up by telemeter, but i do agree  that it may be too big of a hammer in general.  We can definitely start with a lighter touch (something on the upgrade UI page, and docs that tell the admin to check the status before starting an upgrade.  I think we already have docs that tell admins to upgrade all their OLM operators before starting an upgrade)

Reply:
Isn't this a bit ugly for the EUS->EUS case? So every intermediate upgrade will need to stop, check the UI to see if there is anything that's showing the warning, then resume assuming nothing is wrong? Or could we potentially have something more explicit in the case of EUS->EUS?

Reply:
There is no "EUS->EUS" upgrade, it's just a series of upgrades from 4.6->4.70->4.8->4.9->4.10. 

So the EUS->EUS experience is already "make sure you're ready to upgrade.  run an upgrade to the next minor.  make sure it succeeded. make sure you're ready for the next tone.  run the next minor.  rinse/repeat until you get to 4.10".

So while this adds an extra step in what it means to make sure you're ready to upgrade, it's not fundamentally changing the fact that EUS->EUS is still a series of independent upgrades that have to be managed properly by an admin.  

What we (as an org) have been asked to do in support of it is:
1) do everything we can to ensure those upgrades will be successful/inform the admins when the upgrade may not be successful
2) do everything we can to ensure those upgrades are smooth/fast were possible